CBP-1289 - Add URLs, disable signiture verification, and fixes

kwhetstone opened the pull request.

AsafRollout commented on src/core.c:
```diff
@@ -277,10 +278,15 @@ ROX_INTERNAL RoxStateCode rox_core_setup(
                 ROX_ERROR("Invalid rollout apikey - must be specified");
                 return RoxErrorEmptyApiKey;
-            } else if (!str_matches(api_key, mongoIdPattern, PCRE2_CASELESS) && !str_matches(api_key, mongoIdPattern, PCRE2_CASELESS)) {
+            } else if (!str_matches(api_key, mongoIdPattern, PCRE2_CASELESS) && !str_matches(api_key, uuidIdPattern, PCRE2_CASELESS)) {
```
using `str_matches(api_key, uuidIdPattern, PCRE2_CASELESS)` twice
can reuse

kwhetstone replied:
I can throw it into a method.  I figured this was temporary anyway as eventually this would be refactored to just being a check for the platform key.

AsafRollout replied:
can just put it in a variable, no need for a method, yet.. 😄

ankurv02 replied:
it's already updated to "platformMatch" bool.

AsafRollout commented on src/core.c:
```diff
@@ -277,10 +278,15 @@ ROX_INTERNAL RoxStateCode rox_core_setup(
             if (!api_key || !api_key[0]) {
                 ROX_ERROR("Invalid rollout apikey - must be specified");
                 return RoxErrorEmptyApiKey;
-            } else if (!str_matches(api_key, mongoIdPattern, PCRE2_CASELESS) && !str_matches(api_key, mongoIdPattern, PCRE2_CASELESS)) {
+            } else if (!str_matches(api_key, mongoIdPattern, PCRE2_CASELESS) && !str_matches(api_key, uuidIdPattern, PCRE2_CASELESS)) {
                 ROX_ERROR("Illegal rollout apikey");
                 return RoxErrorInvalidApiKey;
             }
+            // if this is a platform key, disable signature verification
+            if (str_matches(api_key, uuidIdPattern, PCRE2_CASELESS)) {
+                rox_options_set_disable_signature_verification(rox_options, true);
+                disableSignature = true;
+            }
```
the code should go to Rollout if it's mongoDB id, or to PLATFORM if it's uuid
couldn't see where this is happening

kwhetstone replied:
Digging more into how this is implemented (uses environment variables), I think this needs to be revisited to be more dynamic period.

ankurv02 replied:
I think we are good here as "platformMatch" is checking for UUID.

AsafRollout commented on src/core/consts.c:
```diff
@@ -39,9 +41,12 @@ ROX_INTERNAL size_t _rox_env_return_value_using_mode_env(
-    return str_copy_value_to_buffer(buffer, buffer_size, prod_mode_value);
+    // Default to the platform
+    return str_copy_value_to_buffer(buffer, buffer_size, platform_mode_value);
 }
```
we went away from env var
other SDKs have in RoxOptions a configuration or network configuration property, which includes all URLs
guess this SDK was never updated, this will mean no support for VPC, and can't run it in PLATFORM preprod..
probably need to ask Glen if it should be added.. or just not support it

kwhetstone replied:
Could we make the environment variable a new PR?  This one is pretty large already, and I'd rather do it in a new set of work.  Another positive is that's it's like a new release when we merge so it's not a large lift in terms of getting a new feature out there.

AsafRollout replied:
if the default is back at Rollout-production we can continue other things in a different PR
we can't set the default to be PLATFORM right now IMO

i'm also not sure how versions will work here, i will imagine that it just the latest one, or need to specify a commit, so better change the version as part of this PR as well

kwhetstone replied:
I'll switch the default to be rollout-production which would negate the changes to a lot of the tests.

There aren't versions for this SDK unfortunately; our install directions are to basically clone the directory, run the install script for the SDK, and to go forward from there.  It doesn't seem like the best user experience, but I don't have any experience with developing server-side products in c.

AsafRollout replied:
think we need to discuss with @glen-polo or @elcloudbees about the ROI for supporting this SDK on CBP right now (before anybody is using it)

AsafRollout reviewed (commented):
* date time custom property needs `tsToNum` operator to really do something with it
* missing serialization changes when it's the platform
* for MAU and other purposes, we send the real platform in get config request for client side sdk, i would imagine that c/c++ are for clients 🤷 

(details of features: https://cloudbees.atlassian.net/browse/CBP-626?focusedCommentId=989789)

kwhetstone commented:
Updated:
* added tsToNum function and some tests
* set default environment to PRODUCTION
* set environment variable to PLATFORM if we've identified a uuid -> we might address VPC later

The c/c++ is a server side sdk (I found an error in the example "getting started code"!)  And I found that there's a "realPlatform" property already set in the config.

AsafRollout commented:
can you please add `tsToNum` evaluation tests? (ex https://github.com/rollout/rox-c/blob/15d340ad09c45e3f4a384d4cb81d72207ab70879/tests/eval/test_parser.c#L440)

if we send datetime custom props to Rollout it might fail everything (or just ignore it 🤷 ), i think the serialization changes are very recommended (ex in .net:
flags
https://github.com/rollout/rox-dotnet-sdk/blob/master/rox-core/X-Pack/Network/StateSender.cs#L95-L97
custom props
https://github.com/rollout/rox-dotnet-sdk/blob/master/rox-core/X-Pack/Network/StateSender.cs#L107-L110
)

AsafRollout commented on src/core/consts.c:
```diff
@@ -91,7 +100,8 @@ ROX_INTERNAL size_t rox_env_get_analytics_path(char *buffer, size_t buffer_size)
             "https://qaanalytic.rollout.io",
-            "https://analytic.rollout.io");
+            "https://analytic.rollout.io",
+            "https://fm-analytics.cloudbees.io/impression");
```
did you check impression?
i think the code adds the `/impression` part
(so it should probably be "https://fm-analytics.cloudbees.io" like the other urls)

AsafRollout commented on src/xpack/network.c:
```diff
@@ -152,6 +153,25 @@ static char *_state_sender_serialize_feature_flags(StateSender *sender) {
 }
 
+// skip datetime for non-platform
+static bool _serialize_the_given_property(CustomProperty *property) {
+    bool to_val = true;
+    const char *value;
+    if ((value = getenv(ROX_ENV_MODE_KEY))) {
+        if (!str_equals(value, ROX_ENV_MODE_PLATFORM)) {
+            CustomPropertyType *propType = custom_property_get_type(property);
+            if (propType == &ROX_CUSTOM_PROPERTY_TYPE_DATETIME) {
+                // skip DateTime
+                to_val = false;
+            }
+            // free the property post check no matter what
+            custom_property_free(propType);
+        }
+        // return true for platform
+    }
+    return to_val;
+}
+
```
👍 

there's also a change in flags serialization (adding the external type the platform prefer, even though it should probably work without)
if you want to do it, or open a follow up ticket...
